providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        if (m_cost != ARGON2_MAX_MEMORY)
+            return 0;
+
+    ctx->m_cost = m_cost;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_digest_length(ARGON2_CTX *ctx,
+                                            uint32_t outlen)
+{
+    if (outlen < ARGON2_MIN_OUTLEN || outlen > ARGON2_MAX_OUTLEN)
+        return 0;
+
+    ctx->outlen = outlen;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_secret(ARGON2_CTX *ctx,
+                                     const OSSL_PARAM *p)
+{
+    size_t buflen;
+
+    if (p->data == NULL)
+        return 0;
+
+    if (ctx->secret != NULL) {
+        OPENSSL_clear_free(ctx->secret, ctx->secretlen);
+        ctx->secret = NULL;
+        ctx->secretlen = 0U;
+    }
+
+    if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->secret, 0, &buflen))
+        return 0;
+
+    if (buflen+1 < ARGON2_MIN_SECRET+1 || buflen > ARGON2_MAX_SECRET
+            || buflen > UINT32_MAX) {
+        OPENSSL_free(ctx->secret);
+        ctx->secret = NULL;
+        ctx->secretlen = 0U;
+        return 0;
+    }
+
+    ctx->secretlen = (uint32_t) buflen;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_pwd(ARGON2_CTX *ctx, const OSSL_PARAM *p)
+{
+    size_t buflen;
+
+    if (p->data == NULL)
+        return 0;
+
+    if (ctx->pwd != NULL) {
+        OPENSSL_clear_free(ctx->pwd, ctx->pwdlen);
+        ctx->pwd = NULL;
+        ctx->pwdlen = 0U;
+    }
+
+    if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->pwd, 0, &buflen))
+        return 0;
+
+    if (buflen+1 < ARGON2_MIN_PWD_LENGTH+1 || buflen > ARGON2_MAX_PWD_LENGTH
+            || buflen > UINT32_MAX) {
+        OPENSSL_free(ctx->pwd);
+        ctx->pwd = NULL;
+        ctx->pwdlen = 0U;
+        return 0;
+    }
+
+    ctx->pwdlen = (uint32_t) buflen;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_salt(ARGON2_CTX *ctx, const OSSL_PARAM *p)
+{
+    size_t buflen;
+
+    if (p->data == NULL)
+        return 0;
+
+    if (ctx->salt != NULL) {
+        OPENSSL_clear_free(ctx->salt, ctx->saltlen);
+        ctx->salt = NULL;
+        ctx->saltlen = 0U;
+    }
+
+    if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->salt, 0, &buflen))
+        return 0;
+
+    if (buflen < ARGON2_MIN_SALT_LENGTH || buflen > ARGON2_MAX_SALT_LENGTH
+            || buflen > UINT32_MAX) {
+        OPENSSL_free(ctx->salt);
+        ctx->salt = NULL;
+        ctx->saltlen = 0U;
+        return 0;
+    }
+
+    ctx->saltlen = (uint32_t) buflen;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_ad(ARGON2_CTX *ctx, const OSSL_PARAM *p)
+{
+    size_t buflen;
+
+    if (p->data == NULL)
+        return 0;
+
+    if (ctx->ad != NULL) {
+        OPENSSL_clear_free(ctx->ad, ctx->adlen);
+        ctx->ad = NULL;
+        ctx->adlen = 0U;
+    }
+
+    if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->ad, 0, &buflen))
+        return 0;
+
+    if (buflen+1 < ARGON2_MIN_AD_LENGTH+1 || buflen > ARGON2_MAX_AD_LENGTH
+            || buflen > UINT32_MAX) {
```

Comment:
Isnt this sufficient?
```
buflen < ARGON2_MIN_AD_LENGTH
   || buflen > ARGON2_MAX_AD_LENGTH
```

Reply:
Same question applies to cases above also.

Reply:
You're right, that was an artifact from a pre-providers version of the implementation. Removed, thanks!